docs/api/matching-utilities.mdx

```diff
@@ -0,0 +1,257 @@
+
+```ts title="isRejected usage"
+import { isRejected } from '@reduxjs/toolkit'
+import { requestThunk1, requestThunk2 } from '@virtual/matchers'
+
+const isARejectedAction = isRejected(requestThunk1, requestThunk2)
+
+function handleRejectedAction(action: AnyAction) {
+  if (isARejectedAction(action)) {
+    // action is a rejected action dispatched by either `requestThunk1` or `requestThunk2`
+  }
+}
+```
+
+## `isRejectedWithValue`
+
+A higher-order function that returns a type guard function that may be used to check whether an action is a 'rejected' action creator from the `createAsyncThunk` promise lifecycle that was created by [`rejectWithValue`](./createAsyncThunk#handling-thunk-errors).
+
+```ts title="isRejectedWithValue usage"
+import { isRejectedWithValue } from '@reduxjs/toolkit'
+import { requestThunk1, requestThunk2 } from '@virtual/matchers'
+
+const isARejectedWithValueAction = isRejectedWithValue(
+  requestThunk1,
+  requestThunk2
+)
+
+function handleRejectedWithValueAction(action: AnyAction) {
+  if (isARejectedWithValueAction(action)) {
+    // action is a rejected action dispatched by either `requestThunk1` or `requestThunk2`
+    // where rejectWithValue was used
+  }
+}
+```
+
+## Using matchers to reduce code complexity, duplication and boilerplate
+
+When using the `builder` pattern to construct a reducer, we add cases or matchers one at a time. However, by using `isAnyOf` or `isAllOf`,
+we're able to easily use the same matcher for several cases in a type-safe manner.
+
+First, let's examine an unnecessarily complex example:
+
+```ts title="Example without using a matcher utility"
+import {
+  createAsyncThunk,
+  createReducer,
+  PayloadAction
+} from '@reduxjs/toolkit'
+
+interface Data {
+  isInteresting: boolean
+  isSpecial: boolean
+}
+
+interface Special extends Data {
+  isSpecial: true
+}
+
+interface Interesting extends Data {
+  isInteresting: true
+}
+
+function isSpecial(
+  action: PayloadAction<Data>
+): action is PayloadAction<Special> {
+  return action.payload.isSpecial
+}
+
+function isInteresting(
+  action: PayloadAction<Data>
+): action is PayloadAction<Interesting> {
+  return action.payload.isInteresting
+}
+
+interface ExampleState {
+  isSpecial: boolean
+  isInteresting: boolean
+}
+
+const initialState: ExampleState = {
+  isSpecial: false,
+  isInteresting: false
+}
+
+export const isSpecialAndInterestingThunk = createAsyncThunk(
+  'isSpecialAndInterestingThunk',
+  () => {
+    return {
+      isSpecial: true,
+      isInteresting: true
+    }
+  }
+)
+
+// This has unnecessary complexity
+const loadingReducer = createReducer(initialState, builder => {
+  builder.addCase(isSpecialAndInterestingThunk.fulfilled, (state, action) => {
+    if (isSpecial(action)) {
+      state.isSpecial = true
+    } else if (isInteresting(action)) {
```

Comment:
Due to the `else` here, both examples actually behave differently.
In the matcher case, both execute.

Reply:
Updated. I just copy and pasted that from the original and didn't bother to check 🤦